CRM/Dedupe/Merger.php

```diff
@@ -907,10 +956,19 @@ public static function skipMerge($mainId, $otherId, &$migrationInfo, $mode = 'sa
    *   Duplicate contact which would be deleted after merge operation.
    *
    * @return array|bool|int
+   *   'main_loc_block' => Stores all location blocks associated with the 'main' contact
    */
   public static function getRowsElementsAndInfo($mainId, $otherId) {
     $qfZeroBug = 'e8cddb72-a257-11dc-b9cc-0016d3330ee9';
 
+    $listOptions = array(
+      'location_type_id' => CRM_Core_PseudoConstant::get('CRM_Core_DAO_Address', 'location_type_id'),
+      'phone_type_id' => CRM_Core_PseudoConstant::get('CRM_Core_DAO_Phone', 'phone_type_id'),
+      'provider_id' => CRM_Core_PseudoConstant::get('CRM_Core_DAO_IM', 'provider_id'),
+      'website_type_id' => CRM_Core_PseudoConstant::get('CRM_Core_DAO_Website', 'website_type_id'),
+      'gender_id' => CRM_Core_PseudoConstant::get('CRM_Contact_DAO_Contact', 'gender_id'),
+    );
```

Comment:
This list could quickly get out of control - there are hundreds and hundreds of pseudoconstants in CiviCRM. And hard-coded lists like this are something I've been working to get rid of.
Should be pretty easy to eliminate, since the list is mostly tautological (`'location_type_id'` is mapping `get('CRM_Core_DAO_Address', 'location_type_id')`, etc) all you need to know is the entity which the field belongs to and then you can look it up automatically. Some suggestions:

1. If you're looping through entities like "Contact", "Phone" "Im" etc then you can use the api to fetch pseudoconstants like `civicrm_api3('$entity, 'getoptions', array('field' => $fieldName));` and the api will handle mapping entityName to DAOname for you.
2. To avoid any hardcoding (even hardcoding of whether a field has options or not) our schema already knows all about pseudoconstants - so if you lookup the metadata for an entity you can tell if a field has a "pseudoconstant" property.